packages/e2e-test-app-fabric/windows/RNTesterApp-Fabric/RNTesterApp-Fabric.cpp

```diff
@@ -308,17 +310,108 @@ winrt::Windows::Data::Json::JsonObject DumpVisualTree(winrt::Windows::Data::Json
   pCondition->Release();
 
+  return uiaTree;
+}
+
+winrt::Windows::Data::Json::JsonObject PrintVisualTree(winrt::Microsoft::UI::Composition::Visual node) {
+  winrt::Windows::Data::Json::JsonObject result;
+  if (!node.Comment().empty()) {
+    result.Insert(L"Comment", winrt::Windows::Data::Json::JsonValue::CreateStringValue(node.Comment()));
+  }
+  winrt::Windows::Data::Json::JsonArray visualSize;
+  visualSize.Append(winrt::Windows::Data::Json::JsonValue::CreateNumberValue(node.Size().x));
+  visualSize.Append(winrt::Windows::Data::Json::JsonValue::CreateNumberValue(node.Size().y));
+  result.Insert(L"Size", visualSize);
+  winrt::Windows::Data::Json::JsonArray visualOffset;
+  visualOffset.Append(winrt::Windows::Data::Json::JsonValue::CreateNumberValue(node.Offset().x));
+  visualOffset.Append(winrt::Windows::Data::Json::JsonValue::CreateNumberValue(node.Offset().y));
+  visualOffset.Append(winrt::Windows::Data::Json::JsonValue::CreateNumberValue(node.Offset().z));
+  result.Insert(L"Offset", visualOffset);
+  result.Insert(L"Opacity", winrt::Windows::Data::Json::JsonValue::CreateNumberValue(node.Opacity()));
+  auto spriteVisual = node.try_as<winrt::Microsoft::UI::Composition::SpriteVisual>();
+  if (spriteVisual) {
+    result.Insert(L"Visual Type", winrt::Windows::Data::Json::JsonValue::CreateStringValue(L"SpriteVisual"));
+    auto spriteBrush = spriteVisual.Brush();
+    if (spriteBrush) {
+      winrt::Windows::Data::Json::JsonObject brush;
+      auto colorBrush = spriteBrush.try_as<winrt::Microsoft::UI::Composition::CompositionColorBrush>();
+      if (colorBrush) {
+        brush.Insert(L"Brush Type", winrt::Windows::Data::Json::JsonValue::CreateStringValue(L"ColorBrush"));
+        auto colorString = L"rgba(" + winrt::to_hstring(colorBrush.Color().R) + L", " +
+            winrt::to_hstring(colorBrush.Color().G) + L", " + winrt::to_hstring(colorBrush.Color().B) + L", " +
+            winrt::to_hstring(colorBrush.Color().A) + L")";
+        brush.Insert(L"Color", winrt::Windows::Data::Json::JsonValue::CreateStringValue(colorString));
+        result.Insert(L"Brush", brush);
+      }
+    }
+  } else {
+    result.Insert(L"Visual Type", winrt::Windows::Data::Json::JsonValue::CreateStringValue(L"Visual"));
+  }
+  return result;
+}
+
+winrt::Windows::Data::Json::JsonObject DumpVisualTreeRecurse(
+    winrt::Microsoft::UI::Composition::Visual node,
+    winrt::hstring accessibilityId,
+    boolean targetNodeHit) {
+  winrt::Windows::Data::Json::JsonObject result;
+  boolean targetNodeFound = false;
+  if (targetNodeHit) {
+    result = PrintVisualTree(node);
+  }
+
+  auto containerNode = node.try_as<winrt::Microsoft::UI::Composition::ContainerVisual>();
+  if (containerNode == nullptr) {
+    return result;
+  }
+  auto nodeChildren = containerNode.Children();
+  winrt::Windows::Data::Json::JsonArray children;
+  for (auto childVisual : nodeChildren) {
+    if (!targetNodeHit && childVisual.Comment() == accessibilityId) {
```

Comment:
So at present I'm iterating through the Microsoft::UI::Composition::Visual types which only have a small set of properties defined Comment being one of them. If we wanted to define a unique property to save the testId to then I believe I would have to change the algorithm to iterate through the Microsoft::ReactNative::IVisual type which allows us to define new properties. Something to note it looks like all the properties do just get passed into the corresponding composition properties. So it would be a distinct design choice to have the IVisual itself store property data. It should be possible for me to rewrite the tests in the way just depends on which design choice we want and if we think customers will want to override the Comment property to something other than the testId. @acoates-ms any thoughts?

Reply:
Yes the E2E tests are not going to be able to read anything from the MS.RN.Visuals.  It can only access the underlying composition visuals.  You could store the testid as a property within the Visual.Properties() object.

Reply:
@acoates-ms Yea that was another option I was looking at, but it doesn't look like Properties() supports storing string data (https://learn.microsoft.com/en-us/windows/windows-app-sdk/api/winrt/microsoft.ui.composition.compositionpropertyset?view=windows-app-sdk-1.4). There seems to only be the option to store booleans, colors, and floats/vectors of floats.

Reply:
Hmm. maybe comment is the right mapping for this then.  -- To @jonthysell's comment, unlike how we were using Tag, this would only be for components that have set a testId - which is more explicit that the user of the JS wants something to be set on the native side.

Reply:
Cool makes sense to me! @jonthysell sound okay to you?